// amount of peripherals
#define ADC1_CHANNELS 16
#define ADC3_CHANNELS 7

#define PCF_MAX_INSTANCES 2
#define ADC_MAX_INSTANCES 8

#define SERVO_COUNT 18
#define VOUT_COUNT 6


// I2C addresses
#define EXPANDER1_ADDRESS 0x70
#define EXPANDER2_ADDRESS 0x72

#define POT1_ADDRESS 0x5A
#define POT2_ADDRESS 0x5C
#define POT3_ADDRESS 0x58
#define POT4_ADDRESS 0x5E


// ADCs and filters configuration
#define SPI_ADC_COUNT 8
#define SPI_ADC_AVG_SAMPLES 10

#define ADC_AVG_SAMPLES 10


// STM ADCs pin mapping
#define I_SERVO3_ADC1_rank 0
#define I_SERVO4_ADC1_rank 1
#define I_SERVO5_ADC1_rank 2
#define I_SERVO6_ADC1_rank 3
#define I_SERVO7_ADC1_rank 4
#define I_SERVO8_ADC1_rank 5
#define I_SERVO9_ADC1_rank 6
#define I_SERVO10_ADC1_rank 7
#define I_SERVO11_ADC1_rank 8
#define I_SERVO12_ADC1_rank 9
#define I_SERVO13_ADC1_rank 10
#define I_SERVO14_ADC1_rank 11
#define I_SERVO15_ADC1_rank 12
#define I_SERVO16_ADC1_rank 13
#define I_SERVO17_ADC1_rank 14
#define I_SERVO18_ADC1_rank 15

#define I_SERVO1_ADC3_rank 5
#define I_SERVO2_ADC3_rank 6
#define U_TEMP_ADC3_rank 0
#define I_MANIP_SENSE_ADC3_rank 1
#define I_5V_POW_SENSE_ADC3_rank 2
#define I_3V3_POW_SENSE_ADC3_rank 3
#define I_STANDBY_SENSE_ADC3_rank 4


// expanders pin mapping
#define EXPANDER1_CONV1_EN 0
#define EXPANDER1_CONV2_EN 1
#define EXPANDER1_CONV3_EN 2
#define EXPANDER1_CONV4_EN 3
#define EXPANDER1_CONV5_EN 4
#define EXPANDER1_POT_HVC 6
#define EXPANDER1_POT_EN 7

#define ADC1_CS_PIN 0
#define ADC2_CS_PIN 1
#define ADC3_CS_PIN 2
#define ADC4_CS_PIN 3
#define ADC5_CS_PIN 4
#define ADC_I_SUPPLY_CS_PIN 5
#define ADC_U_SUPPLY_CS_PIN 6
#define ADC_U_BAT_CS_PIN 7


// coefficients
#define I_SERVO_COEFF 1.007326f
#define I_MANIP_COEFF 8.058608f
#define I_5V_POW_COEFF 4.029304f
#define I_3V3_POW_COEFF 1.007326f
#define I_STANDBY_COEFF 1.007326f

#define U_CONVERTER1_COEFF 3.643975012f
#define U_CONVERTER2_COEFF 3.620640431f
#define U_CONVERTER3_COEFF 3.62853544f
#define U_CONVERTER4_COEFF 6.223229088f
#define U_CONVERTER5_COEFF 3.539222615f
#define I_SUPPLY_COEFF 15.70131418f
#define U_SUPPLY_COEFF 6.66699749f
#define U_BAT_ADC_COEFF 6.467963767f

#define TEMP_SENSOR_COEFF_A -0.01179f
#define TEMP_SENSOR_COEFF_B 1.8528f


#define CONV1_3_RFB1 1000.0f
#define CONV1_3_RFB2 2700.0f
#define CONV1_3_RFB3 750.0f

#define CONV4_RFB1 560.0f
#define CONV4_RFB2 3300.0f
#define CONV4_RFB3 750.0f

#define POT_RESISTANCE 10000.0f
#define POT_RESISTANCE_OFFSET 125.0f


// constants values and other coefficients
#define CONV_EN_OC_DELAY 50
#define EN_SERVOS_TICK_DELAY 25
#define VOUT_TICK_DELAY 25


// CAN IDs
#define CAN_ID_MIN 96
#define CAN_ID_MAX 159


#define CAN_ID_GET_ERROR_CODES_REQ 0x64
#define CAN_ID_GET_ERROR_CODES_ACK 0x65

#define CAN_ID_TURN_OFF_REQ 0x66
#define CAN_ID_TURN_OFF_ACK 0x67

#define CAN_ID_PRESSED_OFF_BUTTON_ACK 0x68
#define CAN_ID_PRESSED_OFF_BUTTON 0x69

#define CAN_ID_SET_SERVOS_STATES_REQ 0x6A
#define CAN_ID_SET_SERVOS_STATES_ACK 0x6B
#define CAN_ID_SET_MANIP_STATE_REQ 0x6C
#define CAN_ID_SET_MANIP_STATE_ACK 0x6D
#define CAN_ID_SET_VOLTAGE_OUTPUTS_STATES_REQ 0x6E
#define CAN_ID_SET_VOLTAGE_OUTPUTS_STATES_ACK 0x6F

#define CAN_ID_GET_SERVO_CURRENT_1_3_REQ 0x70
#define CAN_ID_GET_SERVO_CURRENT_1_3_ACK 0x71
#define CAN_ID_GET_SERVO_CURRENT_4_6_REQ 0x72
#define CAN_ID_GET_SERVO_CURRENT_4_6_ACK 0x73
#define CAN_ID_GET_SERVO_CURRENT_7_9_REQ 0x74
#define CAN_ID_GET_SERVO_CURRENT_7_9_ACK 0x75
#define CAN_ID_GET_SERVO_CURRENT_10_12_REQ 0x76
#define CAN_ID_GET_SERVO_CURRENT_10_12_ACK 0x77
#define CAN_ID_GET_SERVO_CURRENT_13_15_REQ 0x78
#define CAN_ID_GET_SERVO_CURRENT_13_15_ACK 0x79
#define CAN_ID_GET_SERVO_CURRENT_16_18_REQ 0x7A
#define CAN_ID_GET_SERVO_CURRENT_16_18_ACK 0x7B

#define CAN_ID_GET_I_MANIP_I_5V_POW_I_3V3_POW_REQ 0x7C
#define CAN_ID_GET_I_MANIP_I_5V_POW_I_3V3_POW_ACK 0x7D
#define CAN_ID_GET_I_STANDBY_I_SUPPLY_REQ 0x7E
#define CAN_ID_GET_I_STANDBY_I_SUPPLY_ACK 0x7F

#define CAN_ID_GET_U_CONVETERS_1_3_REQ 0x80
#define CAN_ID_GET_U_CONVETERS_1_3_ACK 0x81
#define CAN_ID_GET_U_CONVETERS_4_5_REQ 0x82
#define CAN_ID_GET_U_CONVETERS_4_5_ACK 0x83
#define CAN_ID_GET_U_SUPPLY_U_BAT_REQ 0x84
#define CAN_ID_GET_U_SUPPLY_U_BAT_ACK 0x85

#define CAN_ID_GET_TEMPERATURES_REQ 0x86
#define CAN_ID_GET_TEMPERATURES_ACK 0x87

#define CAN_ID_SET_CONVERTER1_VOLTAGE_REQ 0x88
#define CAN_ID_SET_CONVERTER1_VOLTAGE_ACK 0x89
#define CAN_ID_SET_CONVERTER2_VOLTAGE_REQ 0x8A
#define CAN_ID_SET_CONVERTER2_VOLTAGE_ACK 0x8B
#define CAN_ID_SET_CONVERTER3_VOLTAGE_REQ 0x8C
#define CAN_ID_SET_CONVERTER3_VOLTAGE_ACK 0x8D
#define CAN_ID_SET_CONVERTER4_VOLTAGE_REQ 0x8E
#define CAN_ID_SET_CONVERTER4_VOLTAGE_ACK 0x8F


// setup parameters
#define CONVERTER1_SETUP_VOLTAGE 7.4f
#define CONVERTER2_SETUP_VOLTAGE 7.4f
#define CONVERTER3_SETUP_VOLTAGE 7.4f
#define CONVERTER4_SETUP_VOLTAGE 12.0f
